pynetworktables2js/js/dashboard-widget.js

```diff
@@ -0,0 +1,181 @@
+			*  value can be retrieved from the robot.
+			*  
+			*  @param widget: The parent widget object
+			*  
+			*  @return: The updated value that's put into the NetworkTable
+			*/
+			updateRobot : function(widget) {}
+			
+		},
+		 
+	    _create : function() {
+	    	
+	    	this._id = nextID++;
+	    	
+	    	// Sets the type of the widget if one was provided
+	    	var type = $.pynetworktables2js.dashboardWidget.types[this.options.type];
+	    	
+	    	if(type) {
+	    		$.extend(true, this.options, type);
+	    	}
+	    	
+	    	// Add slash to front and remove slash from end
+	    	var key = this.options.networkTables.key;
+	    	if(key) {
+	    		
+	    		if(key[0] !== '/')
+	    			key = '/' + key;
+	    		
+	    		if(key[ key.length - 1] === '/')
+	    			key = key.substring(0, key.length - 1);
+	    				
+	    	}
+	    	
+	    	// Gets content from the url if one was provided
+	    	var url = this.options.url;
+	    	
+	    	if(url) {
+	    		
+	    		var widget = this;
+	    		
+	    		$.ajax({
+    			   type: 'GET',
+    			   url: url,
+    			   dataType: 'html',
+    			   success: function(content) {
+    				   widget.element.append(content);
+    			   },
+    			   complete: function() {
+    				   
+    				   var options = widget.options;
+    				   
+    				   options.init(widget);
+    				   registeredWidgets[widget._id] = widget;
+    				   
+    				   if(options.networkTables.fromRobot) {
+    					   var value = NetworkTables.getValue(options.networkTables.key, options.networkTables.defaultValue);
+    					   options.updateWidget(value, false, widget);
+    				   }
+    				   
+    			   }
+    			});
+	    	
+	    	} else {
+	    		var options = this.options;
+	    		options.init(this);
+				registeredWidgets[this._id] = this;
+				
+				if(options.networkTables.fromRobot) {
+					var value = NetworkTables.getValue(options.networkTables.key, options.networkTables.defaultValue);
+					options.updateWidget(value, false, this);
+				}
+	    	}
+	    },
+	    
+	   /**
+	    * When the widget is removed from the DOM remove from registered widgets as well
+	    */
+	    _destroy : function() {
+	    	delete registeredWidgets[this._id];
+	    },
+	    
+	   /**
+	    * Sets options
+	    */
+	    _setOption : function( key, value ) {
+	        this.options[ key ] = value;
+	    },
+		
+		
+	   /**
+	    *  Calling this updates the the NetworkTables. Should only be called if the
+	    *  NetworkTables.fromRobot option is false.
+	    */
+		update : function() {
+			
+	    	var networkTablesOptions = this.options.NetworkTables;
+	    	    	
+	    	if(networkTablesOptions.fromRobot || !networkTablesOptions.key)
+	    		return;
+	    	
+	    	NetworkTables.putValue(networkTablesOptions.key, this.options.updateRobot(this));
+	    	  	
+		}
+	 
+	});
+	
+	
+	// Notify widgets of incoming NetworkTables values
+	NetworkTables.addGlobalListener(function(key, value, is_new) {
```

Comment:
This doesn't really make sense. There's a reason why there's a key listener -- use that instead of writing essentially the same thing.

Additionally, this implementation is bad and could potentially cause a lot of extra overhead -- particularly because as the number of widgets grows, the length of time needed to process each update grows, regardless of whether the widgets are using that key or not.